### PR TITLE
[Documents] Add required option for editables

### LINKED ISSUE #3672: [New Feature] Define Required editables like input or wysiwyg
If the `<h1>` is empty on a Document i can't publish a Document!
```<h1><?= $this->input('headline', [ 'required' => true ]); ?></h1>```

### PR DESCRIPTION
**Pull request by dvesh3**
## Changes in this pull request  
Resolves #3672

```{{ pimcore_input("headerLine", {'width': 540, ```**```'required': true```**}```) }}```
## Additional info 
It prompts for missing values on **"Save & Publish"** of Documents.

Supported Types:
- **Input** 
- **WYSIWYG**
- **Textarea**
- **Numeric**

### CODE REVIEW

File: models/Document/PageSnippet.php [hunk 1]
```diff
@@ -607,4 +619,46 @@ public function getUrl($hostname = null, $scheme = null)
 
         return $url;
     }
+
+    /**
+     * @return int|null
+     */
+    public function isContentRequired()
+    {
+        return $this->contentRequired;
```
Comment (brusch) [inline models/Document/PageSnippet.php]: The property `contentRequired` isn't persisted to the database, as the database column is called `requireContent` so the auto-mapping doesn't work. So we need to unify the naming everywhere, I'd anyway prefer a name like `hasRequiredEditables` or something similar, which is more self-explaining. 

Please change that for all methods, properties, column names, variables, ... Thanks!

File: bundles/AdminBundle/Resources/public/js/pimcore/document/edit.js [hunk 2]
```diff
@@ -391,11 +391,17 @@ pimcore.document.edit = Class.create({
 
             for (var i = 0; i < editables.length; i++) {
                 try {
-                    if (editables[i].getName() && !editables[i].getInherited()) {
-                        editableName = editables[i].getName();
-                        values[editableName] = {};
-                        values[editableName].data = editables[i].getValue();
-                        values[editableName].type = editables[i].getType();
+                    if(only == "validate") {
```
Comment (brusch) [inline bundles/AdminBundle/Resources/public/js/pimcore/document/edit.js]: We can collect all editables wich are configured as required already on initialization here: 
https://github.com/pimcore/pimcore/blob/98f8f80933856fad19bba41678caa976e9562b86/bundles/AdminBundle/Resources/public/js/pimcore/document/edit/startup.js#L146

I'd say we can just reference them in another array, something like that should be sufficient: 
```js
try {
	let editable = getEditable(editableConfigurations[i]);
	editables.push(editable);
	if(editable.isRequired()) {
		requiredEditables.push(editable)
	}
} catch (e) {
```

File: bundles/AdminBundle/Resources/public/js/pimcore/document/page_snippet.js [hunk 3]
```diff
@@ -445,5 +445,64 @@ pimcore.document.page_snippet = Class.create(pimcore.document.document, {
...
+    },
+
+    save : function ($super, task, only) {
+        if(task !== "publish") {
+            this.validateRequiredEditables(true);
+        }
+        $super(task, only);
+    },
+
+    validateRequiredEditables: function (dismissalert) {
```
Comment (brusch) [inline bundles/AdminBundle/Resources/public/js/pimcore/document/page_snippet.js]: `dismissalert` camel case for better readability plz 😉

File: bundles/AdminBundle/Resources/public/js/pimcore/document/tag.js [hunk 4]
```diff
@@ -126,6 +128,27 @@ pimcore.document.tag = Class.create({
             fieldname: this.name
         }
         return context;
+    },
+
+    validateValue: function(value, el, parent, mark) {
```
Comment (brusch) [inline bundles/AdminBundle/Resources/public/js/pimcore/document/tag.js]: Should we call that method `validateRequiredValue` ?

File: bundles/AdminBundle/Resources/public/js/pimcore/document/tags/input.js [hunk 5]
```diff
@@ -90,13 +94,14 @@ pimcore.document.tags.input = Class.create(pimcore.document.tag, {
...
 
+        // @TODO validator is based on \Zend\Json\Expr and does not work with Twig templates, to be removed in v7.0
         if(options["validator"]) {
             this.element.isValid = options["validator"];
             this.validateElement();
         }
     },
 
-    checkValue: function () {
+    checkValue: function (mark) {
```
Comment (brusch) [inline bundles/AdminBundle/Resources/public/js/pimcore/document/tags/input.js]: The `mark` argument seems to be useless, as `validateValue()` is called only when `this.required == true` and in this case `mark` is always `true`, so I'd say we can just call `this.validateValue(value, this.element, this, true);` instead. 

Please also change that for the other editables.

File: models/Document/PageSnippet.php (hunk 1 above)
Comment (dvesh3) [reply models/Document/PageSnippet.php]: changed to requireEditableValues and is consistent now with db.

File: bundles/AdminBundle/Resources/public/js/pimcore/document/tags/input.js (hunk 5 above)
Comment (dvesh3) [reply bundles/AdminBundle/Resources/public/js/pimcore/document/tags/input.js]: We still need mark argument as on initial load checkValue() is called which also checks the required editables but it should not mark element as required(red highlight class).

Comment (brusch) [reply bundles/AdminBundle/Resources/public/js/pimcore/document/tags/input.js]: ok, thanks for clarification

File: models/Document/PageSnippet.php [hunk 6]
```diff
@@ -70,6 +71,11 @@ abstract class PageSnippet extends Model\Document
      */
     protected $contentMasterDocumentId;
 
+    /**
+     * @var null|int
+     */
+    protected $requireEditableValues = null;
```
Comment (brusch) [inline models/Document/PageSnippet.php]: I'd prefer `hasRequiredEditables` as the name for this property, as it indicates whether this document has at least one editable configured as required or not. 
@weisswurstkanone what do you think?

Comment (brusch) [reply models/Document/PageSnippet.php]: In other places we're also using e.g. `validateRequiredEditables` so I guess this would be more consistent.

Comment (dvesh3) [reply models/Document/PageSnippet.php]: @brusch that's not true, actually it states that if a document is missing values for any required editables.

Comment (brusch) [reply models/Document/PageSnippet.php]: @dvesh3 oh dear, my fault, sorry 😊
But I'm still not entirely happy with the name, what do you think about `missingRequiredEditable` or `hasMissingRequiredEditable`?

File: bundles/AdminBundle/Resources/public/js/pimcore/document/document.js [hunk 7]
```diff
@@ -89,6 +89,10 @@ pimcore.document.document = Class.create(pimcore.element.abstract, {
         var saveData = this.getSaveData(only);
 
         if (saveData) {
+            if(this.data.requireEditableValues !== null) {
+                saveData.requireEditableValues = intval(this.data.requireEditableValues);
```
Comment (brusch) [inline bundles/AdminBundle/Resources/public/js/pimcore/document/document.js]: I think we should work with `bool` here and let the DB adapter/setter do the required casing.

File: models/Document/PageSnippet.php [hunk 8]
```diff
@@ -607,4 +619,51 @@ public function getUrl($hostname = null, $scheme = null)
...
+    {
+        return $this->requireEditableValues;
+    }
+
+    /**
+     * @param int|null $requireEditableValues
+     */
+    public function setRequireEditableValues($requireEditableValues)
+    {
+        $this->requireEditableValues = $requireEditableValues;
```
Comment (brusch) [inline models/Document/PageSnippet.php]: Cast to `bool`

File: models/Document/PageSnippet.php [hunk 9]
```diff
@@ -607,4 +619,51 @@ public function getUrl($hostname = null, $scheme = null)
...
+    }
+
+    /**
+     * Validates if there is a missing value for required editable
+     */
+    public function checkRequireEditableValues() {
```
Comment (brusch) [inline models/Document/PageSnippet.php]: Should we make that `protected` as it is for internal use only?

File: bundles/AdminBundle/Resources/public/js/pimcore/document/edit.js [hunk 10]
```diff
@@ -404,6 +404,34 @@ pimcore.document.edit = Class.create({
         catch (e2) {
         }
 
+        return values;
+    },
+
+    getRequiredEditableValues: function () {
```
Comment (brusch) [inline bundles/AdminBundle/Resources/public/js/pimcore/document/edit.js]: In my opinion we should call this method `getEmptyRequiredEditables` as this is what this function is actually doing (it is not returning the values of the required editables).

File: bundles/AdminBundle/Resources/public/js/pimcore/document/edit.js [hunk 11]
```diff
@@ -404,6 +404,34 @@ pimcore.document.edit = Class.create({
...
+        var values = {};
```
Comment (brusch) [inline bundles/AdminBundle/Resources/public/js/pimcore/document/edit.js]: I think it's sufficient to use a simple array here for `values` and just return the names of the empty required editables: 

```js
var values = [];
// ...
values.push(editableName); 
// ...
```

File: models/Document/PageSnippet.php (hunk 6 above)
Comment (dvesh3) [reply models/Document/PageSnippet.php]: @brusch I think `missingRequiredEditable` fits better ;)

File: models/Document/PageSnippet.php [hunk 12]
```diff
@@ -607,4 +619,61 @@ public function getUrl($hostname = null, $scheme = null)
...
+                $document = Document::getById($this->getId());
+                if ($document instanceof self) {
+                    // rendering could fail if the controller/action doesn't exist, in this case we can skip the required check
+                    $tagNames = $tagUsageResolver->getUsedTagnames($document);
+                    foreach ($tagNames as $tagName) {
+                        $tag = $document->getElement($tagName);
+                        if ($tag instanceof Tag && in_array($tag->getType(), $allowedTypes)) {
+                            $documentOptions = $tag->getOptions();
+                            if ($tag->isEmpty() && isset($documentOptions['required']) && $documentOptions['required'] == true) {
+                                $this->setMissingRequiredEditable(true);
```
Comment (brusch) [inline models/Document/PageSnippet.php]: I think we can skip the check here, right?

Comment (dvesh3) [reply models/Document/PageSnippet.php]: yes! done :)

**Pull request merged**